### PR TITLE
ROX-23570: Factor out mailing lists from vulnerability reporting

### PR DESCRIPTION
**Pull request by pedrottimark**
## Description

Prerequisite to reuse form and view for `NotifierConfigurations` (also known as delivery destinations) from Vulnerability Reports in Compliance Reports.

```ts
export type NotifierConfiguration = {
    emailConfig: {
        notifierId: string;
        mailingLists: string[];
        customSubject: string;
        customBody: string;
    };
    notifierName: string;
};
```

Add src/Components/NotifierConfiguration folder following previous refactoring:
* src/Components/EmailNotifier
* src/Components/EmailTemplate

### Changes in EmailTemplate folder

1. Edit EmailTemplateForm.tsx file.
    * Add `isReadOnly` prop to reuse form for view.
    * Add `readOnlyVariant` prop to `TextInput` and `TextArea` elements.
    * Add conditional rendering for Reset links.
2. Edit EmailTemplateModal.tsx file.
    * Use `onChange={null}` to reuse modal for view.

### Changes in NotifierConfiguration folder

1. Create NotifierConfigurationForm.tsx file.
    * Caller provides name of property in formik values via `fieldIdPrefix` prop.
    * Caller updates other form state if needed via optional `onDeleteLastNotifierConfiguration` prop.
        Vulnerability Reports resets the schedule.
    * Component updates formik state via `setFieldValue` prop.
2. Move and rename NotifierConfigurationView.tsx file.
    * Replace `formValues: ReportFormValues` with `notifierConfiguraton: NotifierConfiguration` prop.
    * Replace references to `reportParameters` with optional `renderTemplatePreview` prop for presentation that is specific to vulnerability reports.
    * Replace description list with table to solve alignment problems.
    * Replace generic `Modal` with specific `EmailTemplateModal` element.
3. Move and rename NotifierMailingLists.tsx file.
    * Caller updates form state via `setMailingLists` and `setNotifier` props.
        It was a judgement call for NotifierMailingLists component to delegate its updates to NotifierConfigurationForm componen
    * Separate `updateNotifierList` from `useEffect` hook to solve timing problem (see testing step 1).

### Changes in VulnerabilityReporting folder

Use optional props to separate vulnerability-specific dependencies on form values other than notifier configurations.

1. Edit DeliveryDestinationsForm.tsx file.
    * Factor out reusable code into NotifierConfigurationForm component.
2. Edit ReportReviewForm.tsx file.
3. Delete useEmailTemplateModal.ts file and use conditional rendering instead.
4. Edit ReportJobs.tsx file.
5. Edit ViewVulnReportPage.tsx file.

## Checklist
- [x] Investigated and inspected CI test results
- [x] ~Unit test and regression tests added~

## Testing Performed

1. `yarn lint` in ui
2. `yarn build` in ui and then compute branch - master for the following
    * `wc build/static/js/*.js`
        main.js 0 = 4633606 - 4633606
        total -3731 = 11574067 - 11577798
    * `ls -al build/static/js/*.js | wc`
        files 0 = 172 - 172
3. `yarn start` in ui

### Manual testing

Prerequisite: create at least one collection.

Create e-mail notifiers on the fly.

1. Create a report and create email notifier
    * Verify payload and response of POST /v1/notifiers request
    * See that notifier is selected automatically. Baseline code intended but did not succeed because of timing problem `filter` method call before response to GET /v1/notifiers request.
    * Verify review details.
    * Verify payload and response of POST /v2/reports/configurations request
2. Edit the first report and ditto items 2 and 3 above.
3. Clone the first report and ditto items 2 and 3 above.

ReportReviewForm

Misaligned in description lists before changes
Details_alignment

Aligned in table after changes
ReportReviewForm

ViewVulnReportPage
ViewVulnReportPage

### Integration testing

Uh, oh! No tests. Yet.

### CODE REVIEW

File: ui/apps/platform/src/Components/NotifierConfiguration/NotifierConfigurationForm.tsx [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    TemplatePreviewArgs,
+} from 'Components/EmailTemplate/EmailTemplateModal';
+import FormLabelGroup from 'Components/PatternFly/FormLabelGroup';
+import useIndexKey from 'hooks/useIndexKey';
+import { NotifierIntegrationBase } from 'services/NotifierIntegrationsService';
+import { NotifierConfiguration } from 'services/ReportsService.types';
+
+import NotifierMailingLists from './NotifierMailingLists';
+
+function splitAndTrimMailingListsString(mailingListsString: string): string[] {
```
Comment (bradr5) [inline ui/apps/platform/src/Components/NotifierConfiguration/NotifierConfigurationForm.tsx:17]: I noticed this code: `const mailingListsString = mailingLists.join(', ');` inside of `NotifierMailingLists` is making it to where you can't backspace when trying to get rid of a comma. For example, I wouldn't be able to backspace if my input looked like this `test@mail.com, test2@mail.com, ` and my cursor was at the end.

You _could_ do a check on the inputType here to get fix that issue if you think it's worthwhile. Something like this:
```javascript
function splitAndTrimMailingListsString(
    mailingListsString: string,
    event: React.ChangeEvent<HTMLInputElement>
): string[] {
    let adjustedMailingListsString = mailingListsString;
    if (
        adjustedMailingListsString.endsWith(',') &&
        (event.nativeEvent as InputEvent).inputType === 'deleteContentBackward'
    ) {
        adjustedMailingListsString = adjustedMailingListsString.slice(0, -1);
    }
    return adjustedMailingListsString.split(',').map((email) => email.trim());
}
```
Downside being that you have to typecast the nativeEvent and maybe a bit convoluted

File: ui/apps/platform/src/Components/NotifierConfiguration/NotifierConfigurationForm.tsx [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    onDeleteLastNotifierConfiguration?: () => void;
+    renderTemplatePreview?: (args: TemplatePreviewArgs) => ReactElement;
+    setFieldValue: (fieldId: string, value: unknown) => void;
+};
+
+function NotifierConfigurationForm({
+    customBodyDefault,
+    customSubjectDefault,
+    errors,
+    fieldIdPrefix,
```
Comment (bradr5) [inline ui/apps/platform/src/Components/NotifierConfiguration/NotifierConfigurationForm.tsx:37]: Not a change request but I did have to stare at this for a bit to figure out what it's accomplishing. I think a different name might have clued me in quicker, something like `formikValueKey` or `formikFieldKey`. With this prop being so tightly coupled with formik I don't think it'd be bad to include it in the prop name.

File: ui/apps/platform/src/Components/NotifierConfiguration/NotifierMailingLists.tsx [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+                </Flex>
+            </FormLabelGroup>
+            <FormLabelGroup
+                isRequired
+                label="Distribution list"
+                fieldId={`${fieldIdPrefix}.emailConfig.mailingLists`}
+                helperText="Enter an audience, who will receive the scheduled report. Multiple email addresses can be entered with comma separators."
+                errors={errors}
+            >
+                <TextInput
```
Comment (bradr5) [inline ui/apps/platform/src/Components/NotifierConfiguration/NotifierMailingLists.tsx:137]: I see a console complaint about this not including an id or aria-label

**Comment by bradr5** [comment]
I noticed `Configure report scope` and `Email notifier` both show id's briefly when switching between steps in the wizard. I understand this has to do with a selection being made but the component not yet having the name since it needs to be re-fetched each time the step is visited. There are at least a couple solutions worth thinking about:

1. Storing name in formik – I know we prefer keeping formik state as close to the api as possible but I still see value in separating the two in some cases. There might be a clean way to keep the actual form input values in formik state while also maintaining what's important for the api.

2. Lifting `notifiers` and state up – This might add complexity but it would give us preexisting notifier data as we switch back and forth between steps. We could still update notifier data when adding a new notifier, switching steps, or adding a new delivery destination. React context could be really useful in this case.

Screenshot 2024-05-14 at 8 46 42 PM

**Review by bradr5 — APPROVED** [review]
Really solid pattern. For how difficult this was, I think you managed to make it easy to understand. Hopeful it'll be straightforward to use in compliance!

File: ui/apps/platform/src/Components/NotifierConfiguration/NotifierConfigurationForm.tsx (hunk 2 above)
Comment (pedrottimark) [reply ui/apps/platform/src/Components/NotifierConfiguration/NotifierConfigurationForm.tsx:37]: Is `fieldIdPrefixFormikPatternFly` too verbose?
* Formik: first argument of `setFieldValue` function.
* PatternFly: `fieldId` property.

File: ui/apps/platform/src/Components/NotifierConfiguration/NotifierConfigurationForm.tsx (hunk 1 above)
Comment (sachaudh) [reply ui/apps/platform/src/Components/NotifierConfiguration/NotifierConfigurationForm.tsx:17]: I have a feeling that might have been something that existed when I initially worked on VM Reports. @pedrottimark we can try to track that as a separate bug if you want.

File: ui/apps/platform/src/Components/NotifierConfiguration/NotifierConfigurationForm.tsx (hunk 2 above)
Comment (sachaudh) [reply ui/apps/platform/src/Components/NotifierConfiguration/NotifierConfigurationForm.tsx:37]: @pedrottimark what are the possible values for `fieldIdPrefix`? Maybe that would help us understand what kind of name would be more appropriate. From the comment below it seems like we could have `deliveryDestinations` or `notifierConfigurations`. Is that correct or is it supposed to be more generic?

File: ui/apps/platform/src/Containers/Vulnerabilities/VulnerablityReporting/forms/ReportReviewForm.tsx [hunk 4]
```diff
@@ -46,7 +49,26 @@ function ReportReviewForm({ title, formValues }: ReportReviewFormParams): ReactE
             >
                 <ReportParametersDetails formValues={formValues} />
                 <Divider component="div" className="pf-v5-u-py-md" />
-                <DeliveryDestinationsDetails formValues={formValues} />
+                <NotifierConfigurationView
+                    customBodyDefault={defaultEmailBody}
```
[the remark this answers is not in the document]
Comment (pedrottimark) [reply ui/apps/platform/src/Containers/Vulnerabilities/VulnerablityReporting/forms/ReportReviewForm.tsx:53]: The property names are `customBody` and `customSubject` in payload and response.

File: ui/apps/platform/src/Components/NotifierConfiguration/NotifierConfigurationForm.tsx (hunk 2 above)
Comment (pedrottimark) [reply ui/apps/platform/src/Components/NotifierConfiguration/NotifierConfigurationForm.tsx:37]: Yeah, I will add comments with examples in every component:

In `NotifierConfigurationForm` component
* Vulnerability Reports: `deliveryDestinations`
* Compliance Reports might have an extra level: `report.notifierConfigurations`

Within `map` method to render `NotifierMailingLists` element:
* Vulnerability Reports: `deliveryDestinations[0]` and so on
* Compliance Reports might have an extra level: `eport.notifierConfigurations[0]` and so on

File: ui/apps/platform/src/Components/NotifierConfiguration/NotifierConfigurationForm.tsx (hunk 1 above)
Comment (pedrottimark) [reply ui/apps/platform/src/Components/NotifierConfiguration/NotifierConfigurationForm.tsx:24]: Wow, thank you!

Confirmed as existing problem on staging, so will fix in a follow up contribution.

**Pull request merged**